File: java/src/modal/BookableItem.java
package modal;

import exception.NonBookableException;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

public class BookableItem {
    private String name;
    List<Reservation> reservations;
    public BookableItem(String name){
        this.name=name;
        this.reservations=new ArrayList<>();
    }
    private boolean isAvailable(LocalDate startDate, LocalDate endDate) {
        Reservation newReservation = new Reservation(startDate, endDate);
        for (Reservation reservation : reservations) {

            if (!reservation.isReservationValid(newReservation)) {
                return false;
            }
        }
        return true;
    }
    public  void book(LocalDate startDate,LocalDate endDate) throws NonBookableException{

        if(isAvailable(startDate,endDate)){
            reservations.add(new Reservation(startDate, endDate));
            System.out.println("reservation seucessfuly done");
        }
        else {
            throw  new NonBookableException("you can't make a book on this intervalle of date");
        }
    }


}

File: python/modal/bookable_item.py
from modal.reservation import Reservation

class BookableItem:
    def __init__(self, name):
        self.name = name
        self.reservations = []

    def is_available(self, start_date, end_date):
        new_reservation = Reservation(start_date, end_date)
        for reservation in self.reservations:
            if not reservation.is_valid(new_reservation):
                return False
        return True

    def book(self, start_date, end_date):
        if self.is_available(start_date, end_date):
            self.reservations.append(Reservation(start_date, end_date))
            print("Reservation successfully done.")
        else:
            raise Exception("You can't make a reservation on this date.")
